keps/sig-api-machinery/4355-coordinated-leader-election/README.md

```diff
@@ -393,6 +420,89 @@ annotations anyway, so this isn't strictly needed, but it would reduce writes
 from the coordinated election controller to leases that were claimed by
 component instances not using Coordinated Leader Election
 
+### API
+
+The lease lock API will be extended with a new field for election preference, denoted as an enum for strategies for Coordinated Leader Election.
+
+```go
+
+type CoordinatedLeaseStrategy string
+
+// CoordinatedLeaseStrategy defines the strategy for picking the leader for coordinated leader election.
+const (
+  OldestCompatibilityVersionStrategy CoordinatedStrategy = "OldestCompatibilityVersionStrategy"
+  NewestCompatibilityVersionStrategy CoordinatedStrategy = "NewestCompatibilityVersionStrategy"
+)
+
+type LeaseSpec struct {
+  // Strategy indicates the strategy for picking the leader for coordinated leader election
+  Strategy CoordinatedLeaseStrategy `json:"strategy,omitempty" protobuf:"string,6,opt,name=strategy"`
+
+  // EXISTING FIELDS BELOW
+
+	// holderIdentity contains the identity of the holder of a current lease.
+	// +optional
+	HolderIdentity *string `json:"holderIdentity,omitempty" protobuf:"bytes,1,opt,name=holderIdentity"`
+	// leaseDurationSeconds is a duration that candidates for a lease need
+	// to wait to force acquire it. This is measure against time of last
+	// observed renewTime.
+	// +optional
+	LeaseDurationSeconds *int32 `json:"leaseDurationSeconds,omitempty" protobuf:"varint,2,opt,name=leaseDurationSeconds"`
+	// acquireTime is a time when the current lease was acquired.
+	// +optional
+	AcquireTime *metav1.MicroTime `json:"acquireTime,omitempty" protobuf:"bytes,3,opt,name=acquireTime"`
+	// renewTime is a time when the current holder of a lease has last
+	// updated the lease.
+	// +optional
+	RenewTime *metav1.MicroTime `json:"renewTime,omitempty" protobuf:"bytes,4,opt,name=renewTime"`
+	// leaseTransitions is the number of transitions of a lease between
+	// holders.
+	// +optional
+	LeaseTransitions *int32 `json:"leaseTransitions,omitempty" protobuf:"varint,5,opt,name=leaseTransitions"`
+}
+```
+
+For the identity leases, a new lease will be created
+
+```go
+type IdentityLeaseSpec struct {
```

Comment:
apiserver identity migrates to this? (should that KEP be updated to track that change? When does it happen?)

Reply:
This type confuses me. Why is this even a lease? Nothing is elected here, or is it? Isn't this more a `LeaseCandidacy`?

```
apiVersion: coordination.k8s.io/v1
kind: LeaseCandidacy
metadata:
  labels:
    coordination.k8s.io/binary-version: "1.29"
    coordination.k8s.io/compatibility-version: "1.29"
  name: some-custom-controller-0001A
  namespace: kube-system
spec:
  leaseName: some-custom-controller
  leaseDurationSeconds: 10
  renewTime: "2023-12-05T02:33:08.685777Z"
```

I.e.:
- no `can-lead-lease: kube-system/some-custom-controller` label. Everything is informer-based in the controller. No need for label selectors.
- no candidacy holder
- no acquireTime, it's stored in the lease
- no LeaseTransitions, it's stored in the lease
- binary/comp-version as labels, so it is extensible. I could be convinced of a `spec.properties map[string]string`. That's about taste. Either works.

Reply:
As commented above, do we need the strategy here? And if yes, we have to order strategies in case there are candidates where they differ.

Reply:
> We need a way of identifying the `Lease` that a LeaseCandidacy may lead. Would that be defined through the metadata and spec.leaseName? One pro of having both in one field is that it's easier to filter the set of candidates when doing leader election on the apiserver.

Was talking about having a field instead of a label. The apiserver can use either, does not matter. And an external actor likely will not care either about a label. There won't be tons of LeaseCandidacies per namespace. So it's easy to list all e.g. via kubectl. But no super strong opinion here.

Reply:
> no strong preference, I like these ones as fields so we can validate semver easier. Will do some digging to see how that behavior is with labels vs map.

This would be by the strategy. The strategy for kube-oldest-version will consider certain properties/labels as semver. If you choose a `Custom` strategy and implement your own logic, you can assume any type you like, semver or something else.

Reply:
> Was talking about having a field instead of a label. The apiserver can use either, does not matter. And an external actor likely will not care either about a label. There won't be tons of LeaseCandidacies per namespace. So it's easy to list all e.g. via kubectl. But no super strong opinion here.

using field instead of label sounds good. It's also possible to use field selectors for optimization so what I brought up earlier is a non-issue.

Reply:
Re: field vs. label.   API review is going to dig into this for sure, so it's quite possible whatever we recommend here in the KEP will change..

My $.02: To me the only actual benefit to a field is up-front validation. If we don't validate anything about the versions, we'd need to define how we handle the arbitrary strings very carefully. I'd be inclined to just state what our expectations are up-front by making it a validated field..